engine/access/rpc/backend/backend_test.go

```diff
@@ -461,24 +462,39 @@ func (suite *Suite) TestGetProtocolStateSnapshotByBlockID_NonFinalizedBlocks() {
 		err = state.Extend(ctx, newBlock)
 		suite.Require().NoError(err)
 
-		// since block was not yet finalized AtHeight must return an invalid snapshot
-		suite.state.On("AtHeight", newBlock.Header.Height).Return(invalid.NewSnapshot(realstate.ErrUnknownSnapshotReference))
-		// since block was added to the block tree it must be queryable by block ID
-		suite.state.On("AtBlockID", newBlock.ID()).Return(state.AtBlockID(newBlock.ID()))
-
 		params := suite.defaultBackendParams()
 		params.MaxHeightRange = TEST_MAX_HEIGHT
 
 		backend, err := New(params)
 		suite.Require().NoError(err)
 
-		// query the handler for the snapshot for non finalized block
-		snapshotBytes, err := backend.GetProtocolStateSnapshotByBlockID(ctx, newBlock.ID())
-		suite.Require().Nil(snapshotBytes)
-		suite.Require().Error(err)
-		suite.Require().Equal(status.Errorf(codes.InvalidArgument, "failed to retrieve snapshot for block by height %v",
-			fmt.Errorf("%d: %v", newBlock.Header.Height, "block not finalized")).Error(),
-			err.Error())
+		suite.T().Run("failed to get a valid snapshot", func(t *testing.T) {
+			// since block was not yet finalized AtHeight must return an invalid snapshot
+			suite.state.On("AtHeight", newBlock.Header.Height).Return(invalid.NewSnapshot(realstate.ErrUnknownSnapshotReference))
+			// since block was added to the block tree it must be queryable by block ID
+			suite.state.On("AtBlockID", newBlock.ID()).Return(state.AtBlockID(newBlock.ID()))
+			suite.headers.On("BlockIDByHeight", newBlock.Header.Height).Return(newBlock.ID(), nil)
+
+			// query the handler for the snapshot for non finalized block
+			snapshotBytes, err := backend.GetProtocolStateSnapshotByBlockID(ctx, newBlock.ID())
+			suite.Require().Nil(snapshotBytes)
+			suite.Require().Error(err)
+			suite.Require().Equal(codes.Internal, status.Code(err))
+		})
+
+		suite.T().Run("no block has been finalized yet at the height we're looking", func(t *testing.T) {
+			// since block was not yet finalized AtHeight must return an invalid snapshot
+			suite.state.On("AtHeight", newBlock.Header.Height).Return(invalid.NewSnapshot(realstate.ErrUnknownSnapshotReference))
+			// since block was added to the block tree it must be queryable by block ID
+			suite.state.On("AtBlockID", newBlock.ID()).Return(state.AtBlockID(newBlock.ID()))
+			suite.headers.On("BlockIDByHeight", mock.Anything).Return(flow.ZeroID, storage.ErrNotFound)
```

Comment:
This second test case won't pass when all subtests are run, because both subtests are re-using the same block (and attempting to overwrite previous mock calls with the same inputs, which are ignored). It's better to create a new block for each test case, so there aren't subtle interactions between subtests.